Add config flow to InfluxDB integration

RobBie1221 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add config flow for influxdb integration. Import integration level configuration for influxdb integration into ConfigEntry. Implement load/unload entry functions. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/43598

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

RobBie1221 committed:
Add basic infrastructure for ConfigEntry

RobBie1221 committed:
Implement load and unload

RobBie1221 committed:
Unsubscribe from listener,  check on empty config

RobBie1221 committed:
Fix tests part 1

RobBie1221 committed:
Fix tests part 2

RobBie1221 committed:
Fix return type

RobBie1221 committed:
Add config flow tests

RobBie1221 committed:
Validate influxdb connection on import

RobBie1221 committed:
Add exception logging

RobBie1221 committed:
Path client for import test

RobBie1221 committed:
Update import test

RobBie1221 committed:
Add connection fail test

RobBie1221 committed:
First step minimal config test

RobBie1221 committed:
Update import tests, add host for v1

RobBie1221 committed:
Refactor config in tests

RobBie1221 committed:
Add tests

RobBie1221 committed:
Cleanup generation of url

RobBie1221 committed:
Add codeowner

RobBie1221 committed:
Fix codeowner

RobBie1221 committed:
Restore sensor test

RobBie1221 committed:
Add database or bucket to unique id and title

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    return None
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class InfluxDBConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    """Handle a config flow for InfluxDB."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
```
```suggestion
```

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
+
+    async def async_step_import(self, import_config=None) -> ConfigFlowResult:
```
typing

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+            unique_id = f"{host}_{bucket}"
+            title = f"{bucket} ({host})"
+
+        await self.async_set_unique_id(unique_id)
```
the host is not a good unique id, can we get something specific from the database that is unique?

RobBie1221 replied:
I think we should skip setting unique id (was not aware that was an option). We should prevent creation of duplicate entries alternatively.

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,62 @@
+    return None
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
Should we also add the user step? This way we can properly deprecate the YAML config

RobBie1221 replied:
In a follow-up PR :)

joostlek reviewed (changes requested):
Hey @RobBie1221, I've seen your name around! If you feel like it, feel free to drop me a message on Discord :)

RobBie1221 committed:
Address smaller review comments

RobBie1221 committed:
Refactor shutdown to use async_on_unload

RobBie1221 committed:
Split options and data in ConfigEntry creation

RobBie1221 committed:
Fix test coverage

RobBie1221 committed:
Fix ruff

RobBie1221 commented:
I addressed review comments. I also made a change to creating the config entry, I split the data and options (data that we later want to change via option flow is now stored in the options part of the config entry.

RobBie1221 committed:
Fix ruff

joostlek commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -406,7 +406,7 @@ def close_v2():
 
     # Else it's a V1 client
-    if CONF_SSL_CA_CERT in conf and conf[CONF_VERIFY_SSL]:
+    if conf.get(CONF_SSL_CA_CERT) is not None and conf[CONF_VERIFY_SSL]:
         kwargs[CONF_VERIFY_SSL] = conf[CONF_SSL_CA_CERT]
```
why did we change this btw?

joostlek replied:
in that case
```suggestion
    if (cert := conf.get(CONF_SSL_CA_CERT)) is not None and conf[CONF_VERIFY_SSL]:
        kwargs[CONF_VERIFY_SSL] = cert
```

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,157 @@
+    return None
+
+
+class InfluxDBConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
```
```suggestion
class InfluxDBConfigFlow(ConfigFlow, domain=DOMAIN):
```

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,157 @@
+                and entry.data[CONF_HOST] == host
+                and entry.data.get(CONF_BUCKET) == bucket
+            ):
+                if data is not None and options is not None:
+                    changed = self.hass.config_entries.async_update_entry(
+                        entry,
+                        data={**entry.data, **data},
+                        options={**entry.options, **options},
+                    )
+                    if changed and entry.state in (
+                        config_entries.ConfigEntryState.LOADED,
+                        config_entries.ConfigEntryState.SETUP_RETRY,
+                    ):
+                        self.hass.async_create_task(
+                            self.hass.config_entries.async_reload(entry.entry_id)
+                        )
+                raise AbortFlow("already_configured")
```
this doesn't sound like aborting, rather like updating and aborting. Ideally we don't do that anymore and have users use reconfigure for this (unless I am missing something why we do this)

RobBie1221 replied:
The problem is that with the amount of options, we need quite a complex option flow for reconfigure. With this behavior, it always updates from the yaml.
If we don’t want this intermediate behavior, either this PR increases in size (which I wouldn’t do) or we don’t push to dev but an intermediate branch to push all the changes incrementally to have a full config flow and push that in the end to dev.

joostlek replied:
I think like we spoke before, there are also options that are nearly impossible to properly get into a config flow, so I think a good first step that would be mergable would be to just setup the connection via the UI and do the rest via YAML (+ maybe some simple options).

I kinda want to keep the import and user step in the same PR, mostly because the way you do one, affects the other (schema, required data, etc). So if we would merge this as is, and we have to change stuff, we have a harder time in having a clear overview in why we did what we did.

And nevertheless, I am committed on getting this through, so don't be scared to be hit with a "this is too big" :)

RobBie1221 replied:
Ok, let’s do the import and user step together.

joostlek replied:
Oh and to continue on reconfiguring, I think that is something that we can do quite fast after implementing this one, so the options are still changeable. Everything not required for the connection (the optional things) can be set using the options flow. I think it would be nice to be in here, but I also don't have the full scope of influxdb in my mind, so if that is indeed one beast, I think adding reconfigure and options for a followup with intermediary branch is nice

RobBie1221 committed:
Minor fix

RobBie1221 committed:
Minor fix

RobBie1221 committed:
Remove import update, remove options from configentry

RobBie1221 committed:
Fix sensor tests

RobBie1221 committed:
Simplify sensor tests

RobBie1221 committed:
Fix reading SSL_CA_CERT

RobBie1221 committed:
Return url

RobBie1221 committed:
Fix tests

RobBie1221 committed:
Remove commented out code

RobBie1221 committed:
First version user step

RobBie1221 committed:
Finish v2 user step

RobBie1221 committed:
Finish v1 user step

RobBie1221 committed:
Add strings for v1 flow

RobBie1221 committed:
Fix import test

RobBie1221 committed:
Add initial tests

RobBie1221 committed:
Add ssl certificate tests

RobBie1221 committed:
Add connection error tests

RobBie1221 committed:
Final test tweaking

RobBie1221 commented:
I added the `user_step` as requested. The import is now one-time. New entries can be made via UI. The options are still fully in yaml.

Let me know if we want to push this directly to dev or to an intermediary branch. I did not do docs yet, if we push directly to dev, docs need to be updated.

joostlek commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +481,95 @@ def close_v1():
+    return True
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
```
Let's have async_setup in the top

RobBie1221 replied:
The top of the file? You mean above `async_setup_entry`? Or all the way?

joostlek replied:
above async_setup_entry

joostlek commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +481,95 @@ def close_v1():
+        self,
+        hass: HomeAssistant,
+        entry: InfluxDBConfigEntry,
+        influx,
+        event_to_json,
+        max_tries,
```
it would be nice to add typing here

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,287 @@
+        elif "database not found" in ex.args[0]:
+            errors = {"base": "invalid_database"}
+        elif "authorization failed" in ex.args[0]:
+            errors = {"base": "invalid_auth"}
+        elif "token" in ex.args[0]:
+            errors = {"base": "invalid_auth_v2"}
```
`Invalid organization, bucket or token`

Is that related to auth? Can we pick a better name than invalid_auth_v2?

RobBie1221 replied:
It is somehow related to auth, your token could not have the correct access rights. However, you could also have mistyped the bucket name or organization. Maybe `invalid_config`?

joostlek commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,287 @@
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Step when user initializes a integration."""
+
+        if user_input is not None:
+            api_version = user_input[CONF_API_VERSION]
+
+            if api_version == "1.x":
+                return await self.async_step_configure_v1()
+
+            return await self.async_step_configure_v2()
+
+        list_of_types = ["1.x", "2.x"]
+
+        schema = vol.Schema(
+            {vol.Required(CONF_API_VERSION, default="2.x"): vol.In(list_of_types)}
+        )
+
+        return self.async_show_form(step_id="user", data_schema=schema)
```
You can look at using a menu here. For an example check genuishub

joostlek commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,597 @@
+    assert result["step_id"] == "configure_v1"
+    assert result["errors"] == {}
+
+    with patch(
+        "homeassistant.components.influxdb.async_setup_entry", return_value=True
+    ):
```
make this a common mock, check for `mock_setup_entry`

joostlek commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,597 @@
+        )
+
+    data = {
+        "api_version": "1",
+        "host": config_url["host"],
+        "port": config_url["port"],
+        "username": config_base.get("username"),
+        "password": config_base.get("password"),
+        "database": config_base["database"],
+        "ssl": config_url["ssl"],
+        "path": config_url["path"],
+        "verify_ssl": config_base["verify_ssl"],
```
Can we use our constants here for the keys?

joostlek commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,597 @@
+)
+async def test_setup_connection_error(
+    hass: HomeAssistant,
+    mock_client,
+    config_base,
+    api_version,
+    get_write_api,
+    test_exception,
+    reason,
```
typing?

joostlek commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,597 @@
+    assert result["step_id"] == "user"
+    assert result["errors"] is None
+
+    if api_version == DEFAULT_API_VERSION:
+        api = "1.x"
+    else:
+        api = "2.x"
```
we don't want if statements in the tests

joostlek commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,597 @@
+        config_base,
+    )
+
+    assert result["type"] is FlowResultType.FORM
+    assert result["errors"] == {"base": reason}
```
Let's repatch to make sure it works and test that we are able to finish the flow

joostlek commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,597 @@
+            data=config_base,
+        )
+
+    assert result["type"] == "create_entry"
```
use the enum

joostlek commented on tests/components/influxdb/test_config_flow.py:
I am missing tests where we verify we don't setup the same database twice, both for the user flows and for the import flows.

RobBie1221 replied:
Do we need that? I made it single config entry (which it also was when using yaml)

joostlek replied:
Oh right that's a good point. In that case we need to make sure we still test it, just to avoid side effects.

RobBie1221 committed:
Address review comments

RobBie1221 committed:
Fix tests

RobBie1221 committed:
Tests step 1

RobBie1221 committed:
Tests step 2

RobBie1221 committed:
Add single instance tests

RobBie1221 commented:
Not stale

emontnemery commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -204,7 +205,9 @@ def validate_version_specific_config(conf: dict) -> dict:
 
 
-def _generate_event_to_json(conf: dict) -> Callable[[Event], dict[str, Any] | None]:
+def _generate_event_to_json(
+    conf: dict,
+) -> Callable[[Event], dict[str, Any] | None]:
```
```suggestion
def _generate_event_to_json(conf: dict) -> Callable[[Event], dict[str, Any] | None]:
```

emontnemery commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +481,95 @@ def close_v1():
+    hass_config = await conf_util.async_hass_config_yaml(hass)
+
+    influx_yaml = CONFIG_SCHEMA(hass_config).get(DOMAIN, {})
+    default_yaml: dict[str, Any] = {
```
```suggestion
    default_filter_settings: dict[str, Any] = {
```

emontnemery commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +481,95 @@ def close_v1():
+        ),
+        CONF_DEFAULT_MEASUREMENT: influx_yaml.get(CONF_DEFAULT_MEASUREMENT),
+        CONF_OVERRIDE_MEASUREMENT: influx_yaml.get(CONF_OVERRIDE_MEASUREMENT),
+        CONF_INCLUDE: influx_yaml.get(CONF_INCLUDE, default_yaml),
+        CONF_EXCLUDE: influx_yaml.get(CONF_EXCLUDE, default_yaml),
```
```suggestion
        CONF_INCLUDE: influx_yaml.get(CONF_INCLUDE, default_filter_settings),
        CONF_EXCLUDE: influx_yaml.get(CONF_EXCLUDE, default_filter_settings),
```

emontnemery commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +481,95 @@ def close_v1():
+    """Unload a config entry."""
+    influx_thread = entry.runtime_data
+
+    influx_thread.shutdown()
```
This is not async safe, wrap it in an executor, and await the executor job from an asyncio.timeout context manager to avoid hanging this task forever.

emontnemery commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,275 @@
+
+        schema = INFLUXDB_V1_SCHEMA
+
+        return self.async_show_form(
+            step_id="configure_v1", data_schema=schema, errors=errors
+        )
```
Use `add_suggested_values_to_schema` so the user doesn't need to fill in the entire form again

emontnemery commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,275 @@
+
+        schema = INFLUXDB_V2_SCHEMA
+
+        return self.async_show_form(
+            step_id="configure_v2", data_schema=schema, errors=errors
+        )
```
Use `add_suggested_values_to_schema` so the user doesn't need to fill in the entire form again

emontnemery commented on tests/components/influxdb/test_init.py:
```diff
@@ -82,66 +91,91 @@ def v2_call(body, precision):
+
+    conf_entries = hass.config_entries.async_entries(DOMAIN)
+
+    assert len(conf_entries) == 1
```
Is this meant to test that import is working?

emontnemery commented on tests/components/influxdb/test_init.py:
Add a test showing we won't create a second config entry if a config entry already exists and there is influxdb configuration in yaml.

emontnemery commented on homeassistant/components/influxdb/__init__.py:
We should ask the user to remove configuration keys which are imported from yaml.
For example like this:
```py
CONFIG_SCHEMA = vol.Schema(
    {
        DOMAIN: vol.All(
            cv.deprecated(CONF_API_VERSION),  # Deprecated in HA Core 2025.10
            cv.deprecated(CONF_HOST),   # Deprecated in HA Core 2025.10
            ...
            INFLUX_SCHEMA,
        )
    },
    extra=vol.ALLOW_EXTRA,
)
```

emontnemery reviewed (commented):
I think this looks good, some comments though

RobBie1221 committed:
Fix prettier

RobBie1221 committed:
Address review comment

RobBie1221 committed:
Fix shutdown in unload entry

RobBie1221 committed:
Fix deprecation

RobBie1221 committed:
Add test to verify no duplicate after import

emontnemery commented on homeassistant/components/influxdb/strings.json:
```diff
@@ -0,0 +1,50 @@
+    },
+    "step": {
+      "configure_v1": {
+        "data": {
```
We wan't `data_description` for new keys, please add that.
It's OK to just make the `data_description` same as `data` if a more elaborate description is not needed.

emontnemery reviewed (commented):
LGTM, just some minor comments on the translations.

RobBie1221 committed:
Address review comment in strings.json

erwindouna commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +497,96 @@ def close_v1():
-    setup(hass, config)
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the InfluxDB component."""
+    conf = config.get(DOMAIN)
```
I wonder why you need to check this with a .get? Why so? :)

RobBie1221 commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,714 @@
+FIXTURE_UPLOAD_UUID = "0123456789abcdef0123456789abcdef"
+
+
+@pytest.fixture
```
I can do that, I opted not to since all other fixtures are not in a separate file. Left it open for a separate refactor PR since this one is already quite big.

RobBie1221 committed:
Address review comment

RobBie1221 committed:
Add v3 to strings

justanotherariel commented on homeassistant/components/influxdb/strings.json:
```diff
@@ -0,0 +1,61 @@
+        },
+        "data_description": {
+          "database": "The name of the database.",
+          "ssl_ca_cert": "Part to the SSL Certificate"
```
```suggestion
          "ssl_ca_cert": "Path to the SSL certificate"
```

justanotherariel commented on homeassistant/components/influxdb/__init__.py:
```diff
@@ -478,42 +497,94 @@ def close_v1():
+    except ConnectionError as err:
+        raise ConfigEntryNotReady(err) from err
+
+    influx_thread = hass.data[DOMAIN] = InfluxThread(
```
Why do we store InfluxThread in `hass.data[DOMAIN]` and `entry.runtime_data`? Storing in `hass.data[DOMAIN]` seems unnecessary.

RobBie1221 replied:
That's indeed unnecessary, good catch

RobBie1221 replied:
Apparantly, it's not unnecessary. It's used in `test_init.py`. It's not something very nice and also not something easily replaced with `entry.runtime_data`.

This is code that was already there. I'm happy to refactor, but maybe in a separate PR?

justanotherariel commented on tests/components/influxdb/test_config_flow.py:
```diff
@@ -0,0 +1,714 @@
+        data=config_base,
+    )
+
+    assert result["type"] == FlowResultType.CREATE_ENTRY
```
```suggestion
    assert result["type"] is FlowResultType.CREATE_ENTRY
```

justanotherariel reviewed (changes requested):
A few more minor edits.

RobBie1221 committed:
Address review comments

RobBie1221 committed:
Restore hass data

justanotherariel committed:
Remove hass.data

justanotherariel committed:
Fix sentence casing

justanotherariel committed:
Fix annotations

justanotherariel commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,277 @@
+
+async def _validate_influxdb_connection(
+    hass: HomeAssistant, data: dict[str, Any]
+) -> dict[str, Any]:
```
That's literally what I did.

justanotherariel commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,277 @@
+        if errors:
+            return self.async_abort(reason=errors["base"])
+
+        return self.async_create_entry(title=title, data=data)
```
Not needed - the manifest option handles this for the integration outside of the integration

justanotherariel commented on homeassistant/components/influxdb/strings.json:
```diff
@@ -0,0 +1,61 @@
+  },
+  "config": {
+    "abort": {
+      "already_configured": "[%key:common::config_flow::abort::already_configured_device%]"
```
Neither string is needed as neither is referenced in the integration. @RobBie1221 please remove "already_configured".

RobBie1221 commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,277 @@
+        _LOGGER.exception("Unknown error")
+        errors = {"base": "unknown"}
+
+    return errors
```
This is not valid. Function `get_influx_connection` is called which does a write test. The write test catches the `ApiException` and raises a `ConnectionError`. This is caught here.

RobBie1221 committed:
Address review comment

RobBie1221 committed:
Restore import

RobBie1221 commented on homeassistant/components/influxdb/config_flow.py:
```diff
@@ -0,0 +1,277 @@
+        ),
+        vol.Required(CONF_TOKEN): TextSelector(
+            TextSelectorConfig(
+                type=TextSelectorType.TEXT,
```
I don't necessarily agree with that

justanotherariel replied:
It is good practice and in line with other integrations. Fixed.

justanotherariel committed:
A few more fixes

justanotherariel committed:
Fix pytest paramteter

justanotherariel reviewed (approved):
Awesome work @RobBie1221, and thank you for your patience 🙌
Cheers.
